fake_craft_cli/dispatcher.py

```diff
@@ -0,0 +1,338 @@
+        if sum([global_args[key] for key in ("quiet", "verbose", "trace")]) > 1:
+            raise ArgumentParsingError(
+                "The 'verbose', 'trace' and 'quiet' options are mutually exclusive."
+            )
+        if global_args["quiet"]:
+            emit.set_mode(EmitterMode.QUIET)
+        elif global_args["verbose"]:
+            emit.set_mode(EmitterMode.VERBOSE)
+        elif global_args["trace"]:
+            emit.set_mode(EmitterMode.TRACE)
+        emit.trace(f"Raw pre-parsed sysargs: args={global_args} filtered={filtered_sysargs}")
+
+        # handle requested help through -h/--help options
+        if global_args["help"]:
+            help_text = self._get_requested_help(filtered_sysargs)
+            raise ProvideHelpException(help_text)
+
+        if filtered_sysargs:
+            command = filtered_sysargs[0]
+            cmd_args = filtered_sysargs[1:]
+
+            # handle requested help through implicit "help" command
+            if command == "help":
+                help_text = self._get_requested_help(cmd_args)
+                raise ProvideHelpException(help_text)
+
+            self.command_args = cmd_args
+            try:
+                self.command_class = self.commands[command]
+            except KeyError:
+                msg = "no such command {!r}".format(command)
+                help_text = help_builder.get_usage_message(msg)
+                raise ArgumentParsingError(help_text)
+        else:
+            # no command passed!
+            help_text = self._get_general_help(detailed=False)
+            raise ArgumentParsingError(help_text)
+
+        emit.trace(f"General parsed sysargs: command={ command!r} args={cmd_args}")
+        return global_args
+
+    def run(self):
+        """Really run the command."""
+        return self.loaded_command.run(self.parsed_command_args)
+
+
+class BaseCommand(ABC):
+    """Base class to build charmcraft commands.
+
+    Subclass this to create a new command; the subclass must define the following attributes:
+
+    - name: the identifier in the command line
+    - help_msg: a one line help for user documentation
+    - overview: a longer multi-line text with the whole command description
+
+    Also it may override the following ones to change their default:
+    - common: if it's a common/starter command, which are prioritized in the help (default to
+      False)
+    - needs_config: will ensure a config is provided when executing the command (default to False)
+
+    It also must/can override some methods for the proper command behaviour (see each
+    method's docstring).
+
+    The subclass must be declared in the corresponding section of main.COMMAND_GROUPS,
+    and will receive and store this group on instantiation (if overriding `__init__`, the
+    subclass must pass it through upwards).
+    """
+
+    common = False
+    needs_config = False
+
+    def __init__(self, config):
+        self.config = config
```

Comment:
**OPEN QUESTION**: All commands in all apps will have the config as an attribute? 

This is **very** handy, and Charmcraft usage of config was coded around this, but in other apps it may be "against the grain" of how config is handed there. Maybe the BaseCommand in the lib should NOT have this, and Charmcraft should subclass it to add it.

Thoughts? Thanks!

Reply:
if config could be none, this would be similar to click's concept of "context"